v0.3 release. (#175)
tensorflow/swift: KNOWN_ISSUES.md

# Swift for TensorFlow Known Issues

This is a curated list of known Swift for TensorFlow issues. With every release,
new issues are added and resolved issues are updated.

Please see the [JIRA issue tracker](https://bugs.swift.org/projects/TF/issues)
for a full list of known issues.

## Version 0.3

### TensorFlow Library

> * [ ] `Tensor` advanced indexing and striding are not supported on GPU.

Reviewer: Is there an API link?

Reviewer 2: They are actually supported. What's not supported on GPU are the subscript setters only (i.e., assignment).

## Version 0.2

### Notebook Environments (Colab and Jupyter)

* [ ] When a runtime error occurs or when you interrupt cell execution,
      resources (e.g. host memory, GPU memory) do not get released. This can
      lead to OOMs. ([TF-338](https://bugs.swift.org/browse/TF-338))
  * Workaround: Restart the runtime ("Runtime > Restart Runtime" in the Colab
    menu bar) to release all the resources.
* [ ] If the last statement on a cell evaluates to a struct that was defined in
      the notebook, then you get an error ("use of undeclared type") instead of
      seeing the value of the statement.
      ([TF-125](https://bugs.swift.org/browse/TF-125))
  * Workaround: Wrap the last statement in `print()`.
* [ ] Using extensions to conform a type to a protocol (e.g. `extension MyType:
      MyProtocol { ... }`), often causes duplicate conformance errors.
      ([TF-162](https://bugs.swift.org/browse/TF-162))

>   * Workaround: Go back to the cell where the type is defined, add the

Reviewer 3: ```suggestion
  * Workaround: Add the conformance in the same cell where the type is defined.
```

    conformance to the original declaration, and re-execute the cell.
* [ ] If a cell that declares a type executes twice, then it creates two
      different types with the same name. Mixing these types can lead to
      confusing error messages like "cannot convert value of type 'MyType' to
      expected argument type 'MyType'".

>       ([TF-156](https://bugs.swift.org/browse/TF-156))

Reviewer 3: Add workaround saying: re-run all cells that use the type to update them all to use the new type?

> * [ ] The autocomplete UI needs improvement. It should show types and

Reviewer 3: ```suggestion
* [ ] The autocomplete UI should show types and
```

      documentation. It should position your cursor in the first argument when
      you complete a function call.

### Swift Standard Library Enhancements

* [ ] The
      [`Differentiable`](https://www.tensorflow.org/swift/api_docs/Protocols/Differentiable)
      protocol's `allDifferentiableVariables` requirement should not have a
      setter. Do not use this directly through a generic type with a
      `Differentiable` conformance constraint.
      ([TF-208](https://bugs.swift.org/browse/TF-208))

### Swift for TensorFlow Deep Learning Library

> * [ ] The `Differentiable` protocol's `allDifferentiableVariables` property

Reviewer 3: Dupe (or at least relatively similar) with above. Could we have only one of the two?

Author: Thanks! Deleted the item from `Swift for TensorFlow Deep Learning Library`, preserving just the one under `Swift Standard Library Enhancements`.

Author: An aside: I considered renaming `Swift Standard Library Enhancements` to `Swift Standard Library` (given that this is the known issues doc) but I feel `Swift Standard Library Enhancements` is more appropriate.

      requirement should not have a setter.
      ([TF-208](https://bugs.swift.org/browse/TF-208))
* [ ] The majority of the Keras layers remain to be implemented; help wanted!
      ([swift-apis#54](https://github.com/tensorflow/swift-apis/issues/54))
* [ ] There is no built-in API for parameter sharing.
* [ ] The
      [`Parameter`](https://www.tensorflow.org/swift/api_docs/Classes/Parameter)
      class does not conform to `Differentiable` yet, and is not recommended for
      general use.
* [ ] The compiler errors displayed when a user-defined layer struct fails to
      fully satisfy the requirements of the `Layer` protocol are unclear.

### Automatic Differentiation

* [ ] Differentiation does not yet support functions with control flow.
      ([TF-356](https://bugs.swift.org/browse/TF-356))

> * [ ] Differentiating functions with respect to an `inout` parameter is not yet

Reviewer 3: Is this now fixed? Or not yet?

Reviewer: It is not a bug, but an unsupported feature.

Reviewer: It would be nice to include [a mailing list discussion](https://groups.google.com/a/tensorflow.org/d/msg/swift/a5vaC2GTY_w/YqV_O7DQCAAJ) about this feature.

Author: Thanks, acknowledged! The mailing list discussion will be mentioned in the release notes email.
